sdk/src/main/java/io/opentelemetry/sdk/metrics/export/IntervalMetricReader.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.metrics.export;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.util.concurrent.MoreExecutors;
+import io.opentelemetry.internal.Utils;
+import io.opentelemetry.sdk.metrics.data.MetricData;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import jdk.nashorn.internal.ir.annotations.Immutable;
+
+/**
+ * Wraps a list of {@link MetricProducer}s and automatically reads and exports the metrics every
+ * export interval.
+ *
+ * @since 0.3.0
+ */
+public final class IntervalMetricReader {
+  private final Exporter exporter;
+  private final ScheduledExecutorService scheduler;
+  private final ScheduledFuture<?> scheduledFuture;
+
+  /**
+   * Stops the worker thread by calling {@link Thread#interrupt()}.
+   *
+   * @since 0.3.0
+   */
+  public void shutdown() {
+    scheduler.shutdown();
+    scheduledFuture.cancel(false);
```

Comment:
I believe that shutting down the executor will do this for you, so I don't think we need to keep a copy of the future around.

Reply:
From the javadoc:
> This method does not wait for previously submitted tasks to
     * complete execution.  Use {@link #awaitTermination awaitTermination}
     * to do that.

So I think that if a scheduled task is in progress we have to wait for them.

Reply:
Do you want the shutdown to not return until all the tasks are complete? All the javadoc says is that it returns immediately, and the tasks will complete. If we want to block until all tasks are done, then the canonical way to do that is to 
```
scheduler.shutdown();
schedule.awaitTermination();
```

Otherwise, if you're ok with letting them finish in the background, then just calling `shutdown()` on the executor is enough. No more tasks will be scheduled, and the currently running ones will complete normally.

There is no need to hold on to the future and cancel it explicitly, either way.

Reply:
So, to sum up: What semantics do you want with this method, @bogdandrutu ? 
Three options I see:

1. Return immediately and let everything in process finish?  In this case we just need to call `shutdown()` on the executor service. 
2. Block until everything has been completed? In this case we call `shutdown()` on the executor service, then `awaitTermination()` on same. 
3. Stop everything immediately and cancel in the in-flight requests as aggressively as possible? In this case, we want to `cancel(true)` the future, and `shutdownNow()` on the executor service.  Then, if we want to block until everything is dead, `awaitTermination()` on the executor service.

Reply:
Did some experimentation, and for option 3) above, you don't have to `cancel(true)` the future. You just have to call `shutdownNow()` on the executor, and it will aggressively cancel any in-flight scheduled tasks.